Q1. Understanding Feature Distributions (Iris Dataset)

Context: Before training any ML model, we analyze feature distributions.

Dataset: Load Iris dataset from sklearn

Tasks:

1. Extract feature matrix as NumPy array
2. Compute:
Mean
Median
Standard deviation
Variance (for each feature)
3. Identify:
Which feature has highest variability and why it matters in ML
4. Convert any one feature into shape (n,1) and explain why ML models expect this format

In [25]:
from sklearn.datasets import load_iris
import numpy as np

# Load dataset
iris = load_iris()

#iris.feature_names

dataset = iris.data

#print(dataset.shape)

mean = np.mean(dataset, axis = 0)
median = np.median(dataset, axis = 0)
std = np.std(dataset, axis = 0)
var = np.var(dataset, axis = 0)

print("\nMean : ", mean)
print("Median : ", median)
print("Standard Deviation : ", std)
print("Variance : ", var)
      
max_var_idx = np.argmax(var)
#print(max_var_idx)
feature_names = iris.feature_names
print("Feature with highest variance: ",feature_names[max_var_idx])
print("Highest variance value: ",var[max_var_idx])

feature = dataset[:,0]
print("Original Shape: ",feature.shape)

feature_reshaped = feature.reshape(-1,1)
print("Original ReShape: ",feature_reshaped.shape)

print("Variance means that there will be differentiation in data for learning")
print("ML Models always expects data to be a table(2D input): Sample and Feature")


Mean :  [5.84333333 3.05733333 3.758      1.19933333]
Median :  [5.8  3.   4.35 1.3 ]
Standard Deviation :  [0.82530129 0.43441097 1.75940407 0.75969263]
Variance :  [0.68112222 0.18871289 3.09550267 0.57713289]
Feature with highest variance:  petal length (cm)
Highest variance value:  3.0955026666666674
Original Shape:  (150,)
Original ReShape:  (150, 1)
Variance means that there will be differentiation in data for learning
ML Models always expects data to be a table(2D input): Sample and Feature


Q2. Data Selection for Model Input

Context: Models are trained on selective features, not full raw data.

Dataset: Iris

Tasks:

1. Extract:
First 100 samples
Only last 2 features
2. Use boolean masking:
Select samples where petal length is greater than dataset mean
3. Count selected samples
4. Explain how this relates to feature-based filtering in ML

In [50]:
from sklearn.datasets import load_iris
import numpy as np
import pandas as pd

# Load dataset
iris = load_iris()

dataset = iris.data

print(iris.feature_names)

print("\nExtract: First 100 samples Only last 2 features")
df = dataset[:,-2:]
print(df.shape)

#Print number of rows and first five rows in original dataset
#print(dataset.shape[0])
#print(dataset[:5])

# Extract petal length
petal_length = dataset[:, 2]

# Compute mean
mean_val = np.mean(petal_length)
print("Mean of Petal length: ",mean_val)

# Boolean mask
mask = petal_length > mean_val

# Select samples
filtered = dataset[mask]

print("\nUse boolean masking: Select samples where petal length is greater than dataset mean")
print(filtered[:5])

print("\nCount selected samples: ", mask.sum())

#Print number of rows and first five rows in filtered dataset with mask
#print(filtered.shape[0])
#print(filtered[:5])

print("\nExplain how this relates to feature-based filtering in ML?")
print("Because Data samples are not analyzed with fully raw data, but only a subset of data that satisfies a condition is required")

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']

Extract: First 100 samples Only last 2 features
(150, 2)
Mean of Petal length:  3.7580000000000005

Use boolean masking: Select samples where petal length is greater than dataset mean
[[7.  3.2 4.7 1.4]
 [6.4 3.2 4.5 1.5]
 [6.9 3.1 4.9 1.5]
 [5.5 2.3 4.  1.3]
 [6.5 2.8 4.6 1.5]]

Count selected samples:  93

Explain how this relates to feature-based filtering in ML?
Because Data samples are not analyzed with fully raw data, but only a subset of data that satisfies a condition is required


Q3. Feature Scaling using NumPy (Very Important)

Context: Most ML models require normalized data.

Dataset: Iris (NumPy array)

Tasks:

1. Apply standardization:

$$
X' = \frac{X - \mu}{\sigma}
$$
	​2. 

Verify:
Mean ≈ 0
Std ≈
on

In [13]:
from sklearn.datasets import load_iris
import numpy as np
import pandas as pd

# Load dataset
iris = load_iris()

dataset = iris.data

print(iris.feature_names)

mean = np.mean(dataset, axis = 0)
std = np.std(dataset, axis = 0)

print("Mean Before standardization: ", mean)
print("Standard deviation Before standardization: ", std)

print(dataset[0:5])

normalized = (dataset - mean) / std

print(normalized[0:5])

mean = np.mean(normalized, axis = 0)
std = np.std(normalized, axis = 0)

print("Mean After standardization: ", mean)
print("Standard deviation After standardization: ", std)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
Mean Before standardization:  [5.84333333 3.05733333 3.758      1.19933333]
Standard deviation Before standardization:  [0.82530129 0.43441097 1.75940407 0.75969263]
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]
[[-0.90068117  1.01900435 -1.34022653 -1.3154443 ]
 [-1.14301691 -0.13197948 -1.34022653 -1.3154443 ]
 [-1.38535265  0.32841405 -1.39706395 -1.3154443 ]
 [-1.50652052  0.09821729 -1.2833891  -1.3154443 ]
 [-1.02184904  1.24920112 -1.34022653 -1.3154443 ]]
Mean After standardization:  [-1.69031455e-15 -1.84297022e-15 -1.69864123e-15 -1.40924309e-15]
Standard deviation After standardization:  [1. 1. 1. 1.]


Q4. Dataset Understanding (Titanic Dataset)

Context: First step in ML pipeline is dataset inspection.

Dataset: Titanic (Seaborn)

Tasks:

1. Display:
head(), tail()
info(), describe()
2. Identify:
Missing values per column
Numerical vs categorical features
3. Explain:
Why identifying feature types is important before modeling

In [24]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")

print("print(df.head()")
print(df.head())

print("print(df.tail()")
print(df.tail())

print("print(df.info()")
print(df.info())

print("print(df.describe()")
print(df.describe())

print(df.isnull().sum())

#Numerical columns
num_cols = df.select_dtypes(include=["int64", "float64"]).columns
print(num_cols)

#catogorical columns
cat_cols = df.select_dtypes(include=["object", "category","bool"]).columns
print(cat_cols)

missing_num = df[num_cols].isnull().sum().sort_values(ascending=False)
print("Numerical Features Missing Values:\n", missing_num)

missing_cat = df[cat_cols].isnull().sum().sort_values(ascending=False)
print("Categorical Features Missing Values:\n", missing_cat)

print("\n Understanding the datatype is required for applying correct transformation, helps implement correct missing data handling")

print(df.head()
   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  
print(df.tail()
     survived  pclass     sex   age  sibsp  parch   fare embarked   class  \
886         0       2    male  27.0      0      0  13.00        S  Second   
887         1       1  female  19

Q5. Filtering Data for Business Logic

Context: Real ML problems require filtering relevant populations.

Dataset: Titanic

Tasks:

1. Filter:
Female passengers in 1st class
2. Compute:
Survival rate for this group
3. Compare with overall survival rate
4. Interpret:
What insight can be used in ML feature engineering

In [33]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")

female_first_class = df[(df["sex" ]== "female") & (df["pclass"] == 1)]
print(female_first_class)

survival_rate = female_first_class["survived"].mean()
print("Survival rate for female first class passengers: ", survival_rate)

overall_survival_rate = df["survived"].mean()
print("Survival rate for all passengers: ", overall_survival_rate)

print("The difference in the survival rate indicates the roles of gender and class in predicting the survival rate")

     survived  pclass     sex   age  sibsp  parch      fare embarked  class  \
1           1       1  female  38.0      1      0   71.2833        C  First   
3           1       1  female  35.0      1      0   53.1000        S  First   
11          1       1  female  58.0      0      0   26.5500        S  First   
31          1       1  female   NaN      1      0  146.5208        C  First   
52          1       1  female  49.0      1      0   76.7292        C  First   
..        ...     ...     ...   ...    ...    ...       ...      ...    ...   
856         1       1  female  45.0      1      1  164.8667        S  First   
862         1       1  female  48.0      0      0   25.9292        S  First   
871         1       1  female  47.0      1      1   52.5542        S  First   
879         1       1  female  56.0      0      1   83.1583        C  First   
887         1       1  female  19.0      0      0   30.0000        S  First   

       who  adult_male deck  embark_town alive  alo

Q6. Handling Missing Data (Critical ML Step)

Context: Models cannot handle null values.

Dataset: Titanic

Tasks:

1. Fill missing age using median
2. Fill missing embarked using mode
3. Drop deck column
4. Explain:
Why different strategies are used for different columns

In [43]:

import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")

#Task 1
age_median = df["age"].median()
df["age_updated"] = df["age"].fillna(age_median)
print(df["age_updated"])

#Task 2
embarked_mode = df["embarked"].mode()
df["embarked_updated"] = df["embarked"].fillna(embarked_mode)
print(df["embarked_updated"])

#Task 3
df.drop(columns = ["deck"],inplace = True)

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888    28.0
889    26.0
890    32.0
Name: age_updated, Length: 891, dtype: float64
0      S
1      C
2      S
3      S
4      S
      ..
886    S
887    S
888    S
889    C
890    Q
Name: embarked_updated, Length: 891, dtype: object


Task 4

Why different strategies are used for different columns?
1. For missing age median gives a more stable (central or middle) value than mean
2. For embarked we used mode because it gives the most frequently occuring port, and making a assumption that most passengers would embark in that port
3. Of the total 891 rows, the missing value for deck is 688, which is ~77% of the rows. Hence we choose to drop it  

Q7. Feature Engineering for Model Improvement

Context: Raw data is rarely useful—features must be created.

Dataset: Titanic

Tasks:

1. Create:
family_size = sibsp + parch + 1
2. Create:
is_alone (binary feature)
3. Compute survival rate by is_alone
4. Explain:
Why engineered features improve ML models

In [2]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")

#Task 1
#Create: family_size = sibsp + parch + 1
df["family_size"] = df["sibsp"] + df["parch"] + 1

print(df[["sibsp", "parch", "family_size"]].head())

#Task 2
#Create: is_alone (binary feature)
df["is_alone"] = (df["family_size"] == 1).astype(int)
print(df[["sibsp", "parch", "family_size","is_alone"]].head())

#Task 3
#Compute survival rate by is_alone
print("Survival rate for is_alone passengers: ", df["is_alone"].mean())

#Task 4
#Explain: Why engineered features improve ML models
print("Sometimes data is not represented with clarity for the human mind and machine to capture information.

   sibsp  parch  family_size
0      1      0            2
1      1      0            2
2      0      0            1
3      1      0            2
4      0      0            1
   sibsp  parch  family_size  is_alone
0      1      0            2         0
1      1      0            2         0
2      0      0            1         1
3      1      0            2         0
4      0      0            1         1


Q8. Group-Based Insights (Customer Segmentation Logic)

Context: ML models often depend on aggregated patterns.

Dataset: Tips

Tasks:

1. Compute:
Average tip by day
Total bill by gender
2. Create new feature:
tip_percentage
3. Identify:
Which day has highest tipping behavior
4. Explain:
How this helps in segmentation models

In [27]:
import seaborn as sns
df = sns.load_dataset('tips')

#Task 1
#Compute: Average tip by day Total bill by gender
avg_tip_by_day = df.groupby("day")["tip"].mean()
print("Average tip by day: ", avg_tip_by_day)

#Task 2
#Create new feature: tip_percentage
df["tip_percentage"] = ( df["tip"] * 100 ) / df["total_bill"]
print(df.head())

#Task 3
#Identify: Which day has highest tipping behavior
highest_tip_day = df.groupby("day")["tip_percentage"].max()
print("Most generous or highest percentage paid as tip by day: ", highest_tip_day)

#Task 4
#Explain: How this helps in segmentation models
print("The segmentation helps understand behaviours and patterns. \
In this example, we could identify tipping behaviour by customer bill or day. \
Could help with Marketing promotions etc")

Average tip by day:  day
Thur    2.771452
Fri     2.734737
Sat     2.993103
Sun     3.255132
Name: tip, dtype: float64
   total_bill   tip     sex smoker  day    time  size  tip_percentage
0       16.99  1.01  Female     No  Sun  Dinner     2        5.944673
1       10.34  1.66    Male     No  Sun  Dinner     3       16.054159
2       21.01  3.50    Male     No  Sun  Dinner     3       16.658734
3       23.68  3.31    Male     No  Sun  Dinner     2       13.978041
4       24.59  3.61  Female     No  Sun  Dinner     4       14.680765
Most generous or highest percentage paid as tip by day:  day
Thur    26.631158
Fri     26.348039
Sat     32.573290
Sun     71.034483
Name: tip_percentage, dtype: float64
The segmentation helps understand behaviours and patterns. In this example, we could identify tipping behaviour by customer bill or day. Could help with Marketing promotions etc


/var/folders/bk/hk3ys3650gx0gg_rs59rqd280000gn/T/ipykernel_12453/1165299141.py:6: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  avg_tip_by_day = df.groupby("day")["tip"].mean()
/var/folders/bk/hk3ys3650gx0gg_rs59rqd280000gn/T/ipykernel_12453/1165299141.py:16: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  highest_tip_day = df.groupby("day")["tip_percentage"].max()


Q9. Advanced Filtering + Subsetting

Context: Training data is often filtered based on business rules.

Dataset: Tips

Tasks:

1. Filter:
total_bill > 20 AND tip < 3
2. Select only relevant columns
3. Analyze:
Is this segment under-tipping?
4. Explain:
How such filtering helps anomaly detection

In [47]:
import seaborn as sns
df = sns.load_dataset('tips')

#Task 1
#Filter: total_bill > 20 AND tip < 3
filtered = df[(df["total_bill"] > 20) & (df["tip"] < 3)]
print(filtered.head())

#Task 2
#Select only relevant columns
filtered[["total_bill","tip","sex","day","time"]].head()

#Task 3
#Analyze: Is this segment under-tipping?
filtered["tip_percentage"] = ( filtered["tip"] * 100 ) / filtered["total_bill"]
print("Tip percentage for this segment: ", filtered["tip_percentage"].mean())

filter_20 = df[(df["total_bill"] > 20)]
filter_20["tip_percentage"] = ( filter_20["tip"] * 100 ) / filter_20["total_bill"]
print("Tip percentage for total_bill greater than 20: ", filter_20["tip_percentage"].mean())

df["tip_percentage"] = ( df["tip"] * 100 ) / df["total_bill"]
print("Tip percentage for overall dataset: ", df["tip_percentage"].mean())

print("The tip_percentage is less compared to overall dataset and one that has total_bill > 20 \
     hence this is under-tipping segment")

#Task 4
#Explain: How such filtering helps anomaly detection
print("It can detect deviations from expected behaviour - high total_bill means high tip")

    total_bill   tip     sex smoker  day    time  size
21       20.29  2.75  Female     No  Sat  Dinner     2
33       20.69  2.45  Female     No  Sat  Dinner     4
48       28.55  2.05    Male     No  Sun  Dinner     3
57       26.41  1.50  Female     No  Sat  Dinner     2
68       20.23  2.01    Male     No  Sat  Dinner     2
Tip percentage for this segment:  9.082752781629468
Tip percentage for total_bill greater than 20:  14.145153490478231
Tip percentage for overall dataset:  16.08025817225047
The tip_percentage is less compared to overall dataset and one that has total_bill > 20      hence this is under-tipping segment
It can detect deviations from expected behaviour - high total_bill means high tip


/var/folders/bk/hk3ys3650gx0gg_rs59rqd280000gn/T/ipykernel_12453/1930427885.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered["tip_percentage"] = ( filtered["tip"] * 100 ) / filtered["total_bill"]
/var/folders/bk/hk3ys3650gx0gg_rs59rqd280000gn/T/ipykernel_12453/1930427885.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filter_20["tip_percentage"] = ( filter_20["tip"] * 100 ) / filter_20["total_bill"]


Q10. Data Transformation for Model Input (NumPy)

Context: ML models require structured numeric arrays.

Tasks:

1. Create two arrays of shape (3,2)
2. Perform:
Vertical stacking
Horizontal stacking
3. Reshape into (2,6)
4. Explain:
Why reshaping is required in ML pipelines

In [57]:
import numpy as np

#Task 1
# Create two arrays of shape (3,2)
x = np.array([[1,2],[2,3],[3,4]])
y = np.array([[5,6],[6,7],[7,8]])

print("Shape of first array: ",x.shape)
print(x)

print("Shape of second array: ",y.shape)
print(y)

#Task 2
# Perform: Vertical stacking Horizontal stacking
horizontal = np.hstack((x,y))
print("Horizontal stack: ",horizontal)

vertical = np.vstack((x,y))
print("Vertical stack: ",vertical)

#Task 3
# Reshape into (2,6)
reshaped = vertical.reshape(2,6)
print("Validate Shape after reshape: ",reshaped.shape)

#Task 4
# Explain: Why reshaping is required in ML pipelines
print("Reshape helps clarify smaples from features and combine more features")

Shape of first array:  (3, 2)
[[1 2]
 [2 3]
 [3 4]]
Shape of second array:  (3, 2)
[[5 6]
 [6 7]
 [7 8]]
Horizontal stack:  [[1 2 5 6]
 [2 3 6 7]
 [3 4 7 8]]
Vertical stack:  [[1 2]
 [2 3]
 [3 4]
 [5 6]
 [6 7]
 [7 8]]
Validate Shape after reshape:  (2, 6)


Q11. End-to-End ML Preprocessing Pipeline (Titanic)

Context: Prepare dataset for ML model training

Tasks:

1. Load dataset
2. Handle missing values
3. Encode categorical variables (sex, embarked)
4. Normalize numerical features
5. Convert to NumPy array

Output:

Final feature matrix ready for model training

In [71]:
#loads data
import seaborn as sns
import pandas as pd
import numpy as np

df = sns.load_dataset("titanic")

#handles missing values
df["age"] = df["age"].fillna(df["age"].median())
df["embarked"] = df["embarked"].fillna(df["embarked"].mode())

#encodes categoricals (sex, embarked)
df["sex"] = df["sex"].map({"male": 0, "female": 1})

#one-hot 
df = pd.get_dummies(df, columns=["embarked"], drop_first=True)

#normalizes numerics
#Using StandardScaler from sklearn. Another method is to compute for each feature like it was done in Question 3.

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

num_cols = ["age", "fare", "sibsp", "parch", "pclass"]

df[num_cols] = scaler.fit_transform(df[num_cols])
print(df.head())

#outputs a NumPy array

features = df.drop("survived", axis=1)
target = df["survived"]

features_array = features.to_numpy()
target_array = target.to_numpy()

print(features_array.shape,target_array.shape)
features_array

   survived    pclass  sex       age     sibsp     parch      fare  class  \
0         0  0.827377    0 -0.565736  0.432793 -0.473674 -0.502445  Third   
1         1 -1.566107    1  0.663861  0.432793 -0.473674  0.786845  First   
2         1  0.827377    1 -0.258337 -0.474545 -0.473674 -0.488854  Third   
3         1 -1.566107    1  0.433312  0.432793 -0.473674  0.420730  First   
4         0  0.827377    0  0.433312 -0.474545 -0.473674 -0.486337  Third   

     who  adult_male deck  embark_town alive  alone  embarked_Q  embarked_S  
0    man        True  NaN  Southampton    no  False       False        True  
1  woman       False    C    Cherbourg   yes  False       False       False  
2  woman       False  NaN  Southampton   yes   True       False        True  
3  woman       False    C  Southampton   yes  False       False        True  
4    man        True  NaN  Southampton    no   True       False        True  
(891, 15) (891,)


array([[0.8273772438659699, 0, -0.5657364610748746, ..., False, False,
        True],
       [-1.5661069258891576, 1, 0.6638610320657843, ..., False, False,
        False],
       [0.8273772438659699, 1, -0.2583370877897099, ..., True, False,
        True],
       ...,
       [0.8273772438659699, 1, -0.10463740114712752, ..., False, False,
        True],
       [-1.5661069258891576, 0, -0.2583370877897099, ..., True, False,
        False],
       [0.8273772438659699, 0, 0.20276197213803718, ..., True, True,
        False]], shape=(891, 15), dtype=object)

Q12. Feature Selection using Statistical Understanding

Context: Not all features improve model performance.

Dataset: Iris

Tasks:

1. Compute correlation matrix
2. Identify highly correlated features
3. Drop redundant features
4. Explain:
Impact on overfitting and model performance

In [78]:
import seaborn as sns
import pandas as pd
import numpy as np

iris = sns.load_dataset("iris")

#Compute correlation matrix
corr_matrix = iris.corr(numeric_only=True)
print(corr_matrix)

#Identify highly correlated features
high_corr = corr_matrix.unstack().sort_values(ascending=False)
high_corr = high_corr[(high_corr > 0.8) & (high_corr < 1.0)]
print("Highly Correlated Features:\n", high_corr)

#Drop redundant features
iris_reduced = iris.drop(columns=["petal_width"])
iris_reduced
#Explain: Impact on overfitting and model performance
print("Impact on overfitting: Having it in the feature set is going to create duplicates and not change the outcome in any significant way")

              sepal_length  sepal_width  petal_length  petal_width
sepal_length      1.000000    -0.117570      0.871754     0.817941
sepal_width      -0.117570     1.000000     -0.428440    -0.366126
petal_length      0.871754    -0.428440      1.000000     0.962865
petal_width       0.817941    -0.366126      0.962865     1.000000
Highly Correlated Features:
 petal_length  petal_width     0.962865
petal_width   petal_length    0.962865
sepal_length  petal_length    0.871754
petal_length  sepal_length    0.871754
sepal_length  petal_width     0.817941
petal_width   sepal_length    0.817941
dtype: float64
having it in the feature set is going to create duplicates and not change the outcome in any significant way


Q13. Custom Feature Transformation using apply()

Context: Real ML requires custom logic transformation.

Dataset: Titanic

Tasks:

1. Create age_group:
Child / Adult / Senior
2. Use .apply()
3. Compute survival rate per group
4. Interpret:
Which segment has highest survival likelihood

In [84]:
import seaborn as sns
import pandas as pd

df = sns.load_dataset("titanic")

#Create age_group: Child / Adult / Senior
#Use .apply()

def age_group(age):
    if pd.isnull(age):
        return "Unknown"
    elif age < 16:
        return "Child"
    elif age <= 60:
        return "Adult"
    else:
        return "Senior"

df["age_group"] = df["age"].apply(age_group)

#Compute survival rate per group
survival_by_age_group = df.groupby("age_group")["survived"].mean()
print(survival_by_age_group)

#Interpret: Which segment has highest survival likelihood
print(survival_by_age_group.idxmax(),survival_by_age_group.max())

age_group
Adult      0.387521
Child      0.590361
Senior     0.227273
Unknown    0.293785
Name: survived, dtype: float64
Child 0.5903614457831325


Q14. Error Analysis Simulation (Model Evaluation Thinking)

Context: Understanding prediction errors is key in ML.

Tasks:

1. Generate 1000 random error values using NumPy
2. Compute:
Mean
Standard deviation
3. Identify outliers:
Using mean ± 2*std
4. Explain:
What high variance means in model performance

In [90]:
import numpy as np

#Generate 1000 random error values using NumPy
rng = np.random.default_rng()
errors = rng.normal(loc=0.0, scale=1.0, size=1000)

#Compute: Mean Standard deviation
mean = np.mean(errors)
std = np.std(errors)

print("Mean: ", mean)
print("Standard Deviation: ", std)

#Identify outliers: Using mean ± 2*std
low = mean - 2 * std
high = mean + 2 * std

outliers = errors[(errors < low) | (errors > high)]

print("Number of Outliers:", len(outliers))
print("Outliers:", outliers)

#Explain: What high variance means in model performance
print("It is difficult to generalize high variance data")

Mean:  0.04046752717179184
Standard Deviation:  1.0282225986192688
Number of Outliers: 39
Outliers: [ 3.10702289 -2.50028941  2.40568093  2.39351358  2.15069955 -2.09135934
  2.55747636 -2.19272083  3.34373533 -2.30983044  2.34644683  2.21789663
 -2.54530121 -2.11714953 -2.6339843  -2.73465355 -2.07656862  2.1764011
 -2.26963631 -2.13505988  2.16265092  3.40971754  2.41357119 -2.15148793
 -2.05009455 -2.11174121 -2.6422707   2.67569532 -2.780066    2.66658387
  2.1969842  -2.87259474  3.03818336  2.25417604 -2.23364019  2.14118092
  2.25380803 -2.55539875 -2.3861915 ]
It is difficult to generalize high variance data


Q15. Business + ML Insight Problem (Tips Dataset)

Context: Convert data analysis into decision-making

Tasks:

1. Create:
tip_percentage
2. Group by:
day and time
3. Identify:
Highest revenue segment
4. Provide:
Recommendation for business strategy
How ML model can use this insight

In [96]:
import seaborn as sns
import pandas as pd

tips = sns.load_dataset("tips")

#Create: tip_percentage
tips["tip_percentage"] = tips["tip"] / tips["total_bill"]

#Percentage could also be calculated as tips["tip_percentage"] = (tips["tip"] * 100) / tips["total_bill"]

#Group by: day and time
revenue_day_time = tips.groupby(["day","time"])["total_bill"].sum().sort_values(ascending=False)

#Identify: Highest revenue segment
print("Highest revenue segment: ", revenue_day_time.head(1))

#Provide: Recommendation for business strategy How ML model can use this insight
print("There are a number of things that can be done - Offering fine dine options for premium customers, \
pricing strategy, revenue predictions, inventory planning, staff planning etc")

Highest revenue segment:  day  time  
Sat  Dinner    1778.4
Name: total_bill, dtype: float64
There are a number of things that can be done - Offering fine dine options for premium customers, pricing strategy, revenue predictions, inventory planning, staff planning etc


/var/folders/bk/hk3ys3650gx0gg_rs59rqd280000gn/T/ipykernel_12453/382914062.py:12: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  revenue_day_time = tips.groupby(["day","time"])["total_bill"].sum().sort_values(ascending=False)
